src/routes/v2/images.rs

```diff
@@ -0,0 +1,187 @@
+                data.context
+            )));
+        }
+        let scopes = vec![context.relevant_scope()];
+
+        let cdn_url = dotenvy::var("CDN_URL")?;
+        let user = get_user_from_headers(&req, &**pool, &redis, &session_queue, Some(&scopes))
+            .await?
+            .1;
+
+        let bytes =
+            read_from_payload(&mut payload, 1_048_576, "Icons must be smaller than 1MiB").await?;
+
+        let hash = sha1::Sha1::from(&bytes).hexdigest();
+        let upload_data = file_host
+            .upload_file(
+                content_type,
+                &format!("data/cached_images/{}.{}", hash, data.ext),
+                bytes.freeze(),
+            )
+            .await?;
+
+        let mut transaction = pool.begin().await?;
+
+        let db_image: database::models::Image = database::models::Image {
+            id: database::models::generate_image_id(&mut transaction).await?,
+            url: format!("{}/{}", cdn_url, upload_data.file_name),
+            size: upload_data.content_length as u64,
+            created: chrono::Utc::now(),
+            owner_id: database::models::UserId::from(user.id),
+            context,
+        };
+
+        // Insert
+        db_image.insert(&mut transaction).await?;
+
+        let image = Image::from(db_image);
+
+        transaction.commit().await?;
+
+        Ok(HttpResponse::Ok().json(image))
+    } else {
+        Err(ApiError::InvalidInput(
+            "The specified file is not an image!".to_string(),
+        ))
+    }
+}
+
+#[derive(Deserialize, Serialize)]
+pub struct ImageEdit {
+    pub set_id: String,
+}
+
+/// Associates an image with a project or thread message
+/// If an image has no contexts associated with it, it may be deleted
+#[patch("{id}")]
+pub async fn image_edit(
+    req: HttpRequest,
+    info: web::Path<(String,)>,
+    web::Query(edit): web::Query<ImageEdit>,
+    pool: web::Data<PgPool>,
+    redis: web::Data<deadpool_redis::Pool>,
+    session_queue: web::Data<AuthQueue>,
+) -> Result<HttpResponse, ApiError> {
+    let string: String = info.into_inner().0;
+    let image_data = database::models::Image::get(&string, &**pool, &redis).await?;
+
+    if let Some(data) = image_data {
+        let mut transaction = pool.begin().await?;
+
+        let scopes = vec![data.context.relevant_scope()];
+        let user = get_user_from_headers(&req, &**pool, &redis, &session_queue, Some(&scopes))
+            .await?
+            .1;
+
+        if user.id == data.owner_id.into() {
+            let checked_id: Option<i64> = match data.context {
+                ImageContext::Project { .. } => {
+                    project_item::Project::get(&edit.set_id, &mut transaction, &redis)
+                        .await?
+                        .map(|x| x.inner.id.0)
+                }
+                ImageContext::Version { .. } => {
+                    let new_id = serde_json::from_str::<ids::VersionId>(&edit.set_id)?;
```

Comment:
not sure any of the two routes are needed. adding is the only route we need IMO, and deleting can just be handled with the detection.

Reply:
resolving with discord convo